File: 2023/1_trebuchet/trebuchet.ex
defmodule Trebuchet do
  # is_digit returns true if 1-9
  def is_digit(char_str) do
    <<ascii::utf8>> = char_str
    ascii > 48 and ascii < 58
  end

  def extract_numbers(line) do
    if line === "" do
      ""
    else
      head = String.first(line)
      tail = String.slice(line, 1, String.length(line))

      numbers = %{
        "one" => "1",
        "two" => "2",
        "three" => "3",
        "four" => "4",
        "five" => "5",
        "six" => "6",
        "seven" => "7",
        "eight" => "8",
        "nine" => "9"
      }

      matches =
        numbers
        |> Enum.map(fn {k, _} -> k end)
        |> Enum.filter(fn number ->
          String.slice(line, 0, String.length(number)) === number
        end)

      cond do
        is_digit(head) ->
          head <> extract_numbers(tail)

        length(matches) === 1 ->
          digit = numbers[Enum.at(matches, 0)]

          digit <> extract_numbers(tail)

        true ->
          extract_numbers(tail)
      end
    end
  end

  def take_ends(numbers) do
    first = String.first(numbers)
    last = String.slice(numbers, String.length(numbers) - 1, String.length(numbers))

    first <> last
  end

  def to_integer(number_str) do
    {integer, _} = Integer.parse(number_str)
    integer
  end

  def main do
    {:ok, contents} = File.read("input.txt")

    contents
    |> String.split("\n", trim: true)
    |> Enum.map(fn line -> extract_numbers(line) end)
    # |> Enum.map(fn line -> IO.inspect(line) end)
    |> Enum.map(fn line -> take_ends(line) end)
    |> Enum.map(fn line -> to_integer(line) end)
    |> Enum.sum()
  end
end

File: 2023/1_trebuchet/trebuchet.py
import sys

numbers = {
  "one": 1,
  "two": 2,
  "three": 3,
  "four": 4,
  "five": 5,
  "six": 6,
  "seven": 7,
  "eight": 8,
  "nine": 9,
}

class FirstAndLastDigits:
  first = -1
  last = -1

  def track_new_digit(self, digit: int) -> None:
    if self.first == -1:
      self.first = digit
    self.last = digit

  def final_number(self) -> int:
    return int(str(self.first) + str(self.last))

def is_digit(char: int) -> bool:
  return char >= 48 and char <= 57

def ascii_to_int(char: int) -> int:
  return char - 48

def matches_word(i: int, line: str) -> [bool, int]:
  for word in numbers:
    if line[i:i+len(word)] == word:
      return [True, numbers[word]]
  return [False, -1]

def calibration_values(file) -> []:
  vals = []

  for line_bytes in file:
    line: str = line_bytes.decode("utf-8")
    tracker = FirstAndLastDigits()

    for i in range(0, len(line)):
      char: int = line_bytes[i]

      if is_digit(char):
        digit = ascii_to_int(char)
        tracker.track_new_digit(digit)
        continue

      [matches, digit] = matches_word(i, line)
      if matches:
        tracker.track_new_digit(digit)
        continue

    vals.append(tracker.final_number())

  return vals

def main():
  with open(sys.argv[1], 'rb') as file:
    vals = calibration_values(file)
    print(vals)
    print(sum(vals))

if __name__ == "__main__":
    # run with `python3 trebuchet.py input2.txt`
    main()
